### PR TITLE
Asr commonvoice recipe update

### PR DESCRIPTION
**Pull request by ftshijt**
Add recipe for more languages of commonvoice dataset
(default add language model training)

- [ ] update README.md

### CODE REVIEW

**Comment by ShigekiKarita** [comment]
`asr2` is reserved for new style recipe e.g., #820. I recommend you to update `egs/commonvoice/asr1`

File: egs/commonvoice/asr1/local/data_prep.pl [hunk 1]
```diff
@@ -46,7 +46,7 @@
   $text =~ tr/a-z/A-Z/;
   print TEXT "$uttId"," ","$text","\n";
   print GNDR "$uttId"," ","$gender","\n";
-  print WAV "$uttId"," sox $db_base/$filepath -t wav -r 16k -b 16 -e signed - |\n";
+  print WAV "$uttId"," ffmpeg -i $db_base/$filepath -f wav -ar 16000 -ab 16 - |\n";
```
Comment (ShigekiKarita) [inline egs/commonvoice/asr1/local/data_prep.pl]: For me, using ffmpeg is fine but why did you update this? I just want to know the reason.

Comment (ftshijt) [reply egs/commonvoice/asr1/local/data_prep.pl]: We found that sox doesn't support MP3 format because of some legal terms.

**Comment by ShigekiKarita** [comment]
Can you also update the dataset description in https://github.com/espnet/espnet/blob/master/egs/README.md

**Comment by ftshijt** [comment]
> Can you also update the dataset description (e.g., lang EN) in https://github.com/espnet/espnet/blob/master/egs/README.md

Sure~

File: egs/commonvoice/asr1/local/data_prep.pl (hunk 1 above)
Comment (sw005320) [reply egs/commonvoice/asr1/local/data_prep.pl]: The default sox does not support mp3 but it can be done with some plug-in install. `ffmpeg` is a robust solution, but it may not be installed as a default tool. So, could you add an installation check of `ffmpeg` like https://github.com/espnet/espnet/blob/83101a26177500542d1fcff9c22bf7752c2f49ad/egs/librispeech/asr1/local/data_prep.sh#L16-L20?

File: egs/commonvoice/asr1/local/filter_text.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import argparse
+import codecs
+from io import open
+import sys
+
+
+PY2 = sys.version_info[0] == 2
```
Comment (sw005320) [inline egs/commonvoice/asr1/local/filter_text.py]: We stopped to support python 2. So you don't need this python 2 related lines.

File: egs/commonvoice/asr1/run.sh [hunk 3]
```diff
@@ -55,23 +53,33 @@ set -e
...
-        local/data_prep.pl ${datadir}/ cv-${part} data/"$(echo ${part} | tr - _)"
+        local/data_prep.pl ${datadir} validated data/"$(echo ${part} | tr - _)"
     done
+
+    # Kaldi Version Split
+    # ./utils/subset_data_dir_tr_cv.sh data/validated data/valid_train data/valid_test_dev
+    # ./utils/subset_data_dir_tr_cv.sh --cv-spk-percent 50 data/valid_test_dev data/valid_test data/valid_dev
+
+    # ESPNet Version
+    # consider duplicated sentences
```
Comment (sw005320) [inline egs/commonvoice/asr1/run.sh]: You can also state that this does not consider the speaker split and this way of split is used for the voxforge recipe.

Comment (ftshijt) [reply egs/commonvoice/asr1/run.sh]: For train-dev-test split,
If we use the official test and dev set and use the rest of the validated set as training set as we discussed this morning, we may encounter the same problem as last week (duplicated word sequence in test and train). That will be a problem for some languages (e.g. welsh) as they will get a very small WER with 10s hours of training data. What do you think about it?

Comment (sw005320) [reply egs/commonvoice/asr1/run.sh]: how about removing the duplicated word sequence appeared in the dev/test set from the train set?

Comment (ftshijt) [reply egs/commonvoice/asr1/run.sh]: It is a possible solution, I will have a try~ Thanks

**Comment by sw005320** [comment]
@YosukeHiguchi, can you give us your comments? This is an extension of your recipe.

**Comment by sw005320** [comment]
Is there speaker information? Is it possible to also exclude test set speakers from the training data (I'm just asking the possibility. You don't have to do it, but this would be better if we still have enough amount of training data).

**Comment by ftshijt** [comment]
> Is there speaker information? Is it possible to also exclude test set speakers from the training data (I'm just asking the possibility. You don't have to do it, but this would be better if we still have enough amount of training data).

There is speaker information. I will test it next.

**Comment by ftshijt** [comment]
> Is there speaker information? Is it possible to also exclude test set speakers from the training data (I'm just asking the possibility. You don't have to do it, but this would be better if we still have enough amount of training data).

**Comment by ftshijt** [comment]
> > Is there speaker information? Is it possible to also exclude test set speakers from the training data (I'm just asking the possibility. You don't have to do it, but this would be better if we still have enough amount of training data).
> 
> There is speaker information. I will test it next.

Unfortunately, it may not be possible to exclude the speakers in the validated set from the test set. I just tested some languages including both large dataset (e.g. de, fr) and small dataset (e.g. cy, tt). The test set has shown to include the majority of the whole speakers (3/5 or more).

**Comment by sw005320** [comment]
OK.
Can you summarize it in a table to record our investigation? (the number of utterances for training/dev/test for each language w/ and w/o considering speaker overlaps)

**Comment by ftshijt** [comment]
> OK.
> Can you summarize it in a table to record our investigation? (the number of utterances for training/dev/test for each language w/ and w/o considering speaker overlaps)


  | SPEAKER |   | UTTERANCE AFTER FILTER
-- | -- | -- | --
LANG | TOTAL_SPEAK | TEST/DEV_SPEAKER | VALIDATE_SPEAKER( /wo test&dev | TOTAL UTTERANCE | UTTERANCE (AFTER SPEAKER FILTER) | UTTERANCE (AFTER TEXT FILTER) | UTTERANCE (AFTER TEXT&SPEECH FILTER)
de | 4852 | 2911 | 1941 | 281207 | 257293 | 58814 | 58814
fr | 2903 | 2466 | 437 | 152362 | 124392 | 77356 | 77356
cy | 725 | 286 | 439 | 34188 | 32362 | 9019 | 9019
tt | 125 | 123 | 2 | 22083 | 15639 | 11859 | 11859
kab | 549 | 455 | 94 | 180582 | 156136 | 53452 | 53452
ca | 1778 | 1543 | 235 | 84818 | 63602 | 36231 | 36127
it | 577 | 549 | 28 | 27062 | 27062 | 9831 | 9831
fa | 1216 | 863 | 353 | 54050 | 47503 | 16341 | 16341
eu | 450 | 365 | 85 | 32962 | 26499 | 11160 | 10926
es | 585 | 546 | 39 | 23763 | 15424 | 6951 | 6930
ru | 61 | 58 | 3 | 19255 | 5718 | 7110 | 2805

**Comment by ftshijt** [comment]
As discussed last Thursday, we use our own test&dev split just like voxforge dataset. If everything goes right, I will start to run for the recipe README

**Comment by ftshijt** [comment]
As for speaker split, it can be found from the previous chart that the speech donation is very imbalance.  For example, the utterances of two speakers in the Tatar(tt) dataset take over around 70% of the total validation data while the other 123 speakers donate the other 30%.

**Comment by sw005320** [comment]
@ftshijt, what is your current status? You don't need to finish all the experiments. I'll merge it if you confirm that it would work with your training/test split.

**Comment by sw005320** [comment]
When you're ready, please edit to remove `[wip]` in the PR title.

**Comment by ftshijt** [comment]
It's ready, I've checked two langs. They are working well

File: egs/commonvoice/asr1/local/filter_text.py (hunk 2 above)
Comment (sw005320) [inline egs/commonvoice/asr1/local/filter_text.py]: Could you reflect on my comments?
As I mentioned, we'd not support python 2 and you can make your code very simple without these PY2 cases.

File: egs/commonvoice/asr1/run.sh [hunk 4]
```diff
@@ -55,23 +53,34 @@ set -e
...
         # use underscore-separated names in data directories.
-        local/data_prep.pl ${datadir}/ cv-${part} data/"$(echo ${part} | tr - _)"
+        local/data_prep.pl ${datadir} ${part} data/"$(echo "${part}_${lang}" | tr - _)"
     done
+
+    # Kaldi Version Split
+    # ./utils/subset_data_dir_tr_cv.sh data/validated data/valid_train data/valid_test_dev
+    # ./utils/subset_data_dir_tr_cv.sh --cv-spk-percent 50 data/valid_test_dev data/valid_test data/valid_dev
+
+    # ESPNet Version (same as voxforge)
```
Comment (sw005320) [inline egs/commonvoice/asr1/run.sh]: This comment is very good.

**Comment by sw005320** [comment]
Thanks!
Also, please add your results to `RESULTS.md` after you get them.
